File: exercism/scala/robot-name/src/main/scala/Robot.scala
import scala.collection.mutable
import scala.util.Random

class Robot() {
  var name = Robot.createUniqueName(2,3)
  def reset(): Unit = {
    Robot.deleteNameFromSet(name)
    name = Robot.createUniqueName(2,3)
  }
}

object Robot {
  val random = new Random()
  val distinctNames = mutable.HashSet.empty[String]

  def createUniqueName(numberOfLetters: Int, numberOfDigit: Int):String = {
    var uniqueName = createRandomName(numberOfLetters,numberOfDigit)
    while(distinctNames contains uniqueName) {
      uniqueName = createRandomName(numberOfLetters,numberOfDigit)
    }
    distinctNames += uniqueName
    uniqueName
  }

  def deleteNameFromSet(name:String):Unit = {
    distinctNames -= name
  }

  def createRandomName(numberOfLetter: Int, numberOfDigit: Int):String = {
    randomString(2) + randomDigits(3)
  }

  def randomDigit(): String = {
    random.nextInt(10).toString
  }

  def randomDigits(times:Int): String = {
    return List.fill(times){randomDigit()}.mkString
  }

  def randomString(length:Int):String = {
    random.shuffle(('A' to 'Z').toList).take(length).mkString
  }

}

File: exercism/java/robot-name/src/main/java/Robot.java
import java.util.HashSet;
import java.util.Random;
import java.util.Set;

public class Robot {

    private static Set<String> existingNames = new HashSet<>();
    private static Random random = new Random();

    private String name = null;

    public String getName() {
        if (name == null) {
            name = generateName();
            while (existingNames.contains(name)) {
                name = generateName();
            }
            existingNames.add(name);
        }
        return name;
    }

    public void reset() {
        this.name = null;
    }

    private String generateName() {
        return new StringBuilder()
                .append(generateRandomUpperCaseLetter())
                .append(generateRandomUpperCaseLetter())
                .append(generateRandomDigit())
                .append(generateRandomDigit())
                .append(generateRandomDigit())
                .toString();
    }

    private char generateRandomUpperCaseLetter() {
        String UPPPERCASE_LETTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        return UPPPERCASE_LETTERS.charAt(random.nextInt(UPPPERCASE_LETTERS.length()));
    }

    private char generateRandomDigit() {
        String DIGITS = "0123456789";
        return DIGITS.charAt(random.nextInt(DIGITS.length()));
    }

}